src/coreclr/tools/aot/crossgen2/Program.cs

```diff
@@ -484,6 +484,15 @@ private int Run(string[] args)
             {
                 var singleCompilationInputFilePaths = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
 
+                var referenceFilePaths = new Dictionary<string, string>();
+                foreach (string key in _referenceFilePaths.Keys)
+                {
+                    referenceFilePaths[key] = _referenceFilePaths[key];
+                }
+                foreach (string key in inputFilePaths.Keys)
+                {
+                    referenceFilePaths[key] = inputFilePaths[key];
```

Comment:
This is not right for `--single-file-compilation` mode. Input files are independent of each other, but they share references. So change with creation of new `var referenceFilePaths` can be fully removed and below it would be:
```c#
typeSystemContext.ReferenceFilePaths = _referenceFilePaths;
```

Reply:
To clarify, you command should look like this. You were not passing all the references.

C:\Work\GitHub\runtime\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root\corerun.exe C:\Work\GitHub\runtime\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root\crossgen2\crossgen2.dll --targetos:windows --targetarch:x64 -O --single-file-compilation --out-near-input -r C:\Work\GitHub\runtime\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root\*.dll -r C:\Work\alpha\Bin.RiderBackend\Rider.Backend.exe C:\Work\alpha\Bin.RiderBackend\*.dll C:\Work\alpha\Bin.RiderBackend\Rider.Backend.exe C:\Work\alpha\Bin.RiderBackend\*.dll

Reply:
We should also pass: `typeSystemContext.InputFilePaths = singleCompilationInputFilePaths;`
in other case we will have NRE:
```
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at ILCompiler.CompilerTypeSystemContext.GetModuleForSimpleName(String simpleName, Boolean throwIfNotFound) in ILCompiler.ReadyToRun.dll:token 0x6000025+0x45
   at Internal.TypeSystem.Ecma.EcmaModule.ResolveAssemblyReference(AssemblyReferenceHandle handle) in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Ecma\EcmaModule.cs:line 588
   at Internal.TypeSystem.Ecma.EcmaModule.EcmaObjectLookupHashtable.CreateValueFromKey(EntityHandle handle) in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Ecma\EcmaModule.cs:line 89
   at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey key) in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Common\Utilities\LockFreeReaderHashtable.cs:line 560
   at Internal.TypeSystem.Ecma.EcmaModule.ResolveTypeReference(TypeReferenceHandle handle) in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Ecma\EcmaModule.cs:line 538
   at Internal.TypeSystem.Ecma.EcmaModule.EcmaObjectLookupHashtable.CreateValueFromKey(EntityHandle handle) in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Ecma\EcmaModule.cs:line 89
   at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey key) in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Common\Utilities\LockFreeReaderHashtable.cs:line 560
   at Internal.TypeSystem.Ecma.EcmaModule.GetObject(EntityHandle handle, NotFoundBehavior notFoundBehavior) in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Ecma\EcmaModule.cs:line 393
   at Internal.TypeSystem.Ecma.EcmaSignatureParser.ResolveHandle(EntityHandle handle) in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Ecma\EcmaSignatureParser.cs:line 85
   at Internal.TypeSystem.Ecma.EcmaSignatureParser.ParseTypeImpl(SignatureTypeCode typeCode) in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Ecma\EcmaSignatureParser.cs:line 0
   at Internal.TypeSystem.Ecma.EcmaSignatureParser.ParseType(SignatureTypeCode typeCode) in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Ecma\EcmaSignatureParser.cs:line 107
   at Internal.TypeSystem.Ecma.EcmaSignatureParser.ParseType() in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Ecma\EcmaSignatureParser.cs:line 329
   at Internal.TypeSystem.Ecma.EcmaSignatureParser.ParseFieldSignature() in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Ecma\EcmaSignatureParser.cs:line 468
   at Internal.TypeSystem.Ecma.EcmaField.InitializeFieldType() in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Ecma\EcmaField.cs:line 98
   at Internal.TypeSystem.MetadataFieldLayoutAlgorithm.ComputeInstanceLayout(DefType defType, InstanceLayoutKind layoutKind) in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Common\MetadataFieldLayoutAlgorithm.cs:line 38
   at Internal.TypeSystem.DefType.ComputeInstanceLayout(InstanceLayoutKind layoutKind) in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Common\DefType.FieldLayout.cs:line 399
   at ILCompiler.CompilerTypeSystemContext.EnsureLoadableTypeUncached(TypeDesc type) in ILCompiler.ReadyToRun.dll:token 0x6000037+0x14c
   at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey key) in C:\Work\GitHub\runtime\src\coreclr\tools\Common\TypeSystem\Common\Utilities\LockFreeReaderHashtable.cs:line 560
   at ILCompiler.CompilerTypeSystemContext.EnsureLoadableMethod(MethodDesc method) in ILCompiler.ReadyToRun.dll:token 0x6000036+0x0
   at ILCompiler.ReadyToRunRootProvider.CheckCanGenerateMethod(MethodDesc method) in ILCompiler.ReadyToRun.dll:token 0x6000107+0x0
   at ILCompiler.ReadyToRunRootProvider.RootMethods(TypeDesc type, String reason, IRootingServiceProvider rootProvider) in ILCompiler.ReadyToRun.dll:token 0x6000106+0x42
   at ILCompiler.ReadyToRunRootProvider.AddCompilationRoots(IRootingServiceProvider rootProvider) in ILCompiler.ReadyToRun.dll:token 0x6000105+0x101
   at ILCompiler.Compilation..ctor(DependencyAnalyzerBase`1 dependencyGraph, NodeFactory nodeFactory, IEnumerable`1 compilationRoots, ILProvider ilProvider, DevirtualizationManager devirtualizationManager, IEnumerable`1 modulesBeingInstrumented, Logger logger, InstructionSetSupport instructionSetSupport) in ILCompile
r.ReadyToRun.dll:token 0x6000120+0x8a
   at ILCompiler.ReadyToRunCodegenCompilation..ctor(DependencyAnalyzerBase`1 dependencyGraph, NodeFactory nodeFactory, IEnumerable`1 roots, ILProvider ilProvider, Logger logger, DevirtualizationManager devirtualizationManager, IEnumerable`1 inputFiles, String compositeRootPath, InstructionSetSupport instructionSetSup
port, Boolean resilient, Boolean generateMapFile, Boolean generateMapCsvFile, Boolean generatePdbFile, Func`2 printReproInstructions, String pdbPath, Boolean generatePerfMapFile, String perfMapPath, Int32 perfMapFormatVersion, Boolean generateProfileFile, Int32 parallelism, ProfileDataManager profileData, ReadyToRunM
ethodLayoutAlgorithm methodLayoutAlgorithm, ReadyToRunFileLayoutAlgorithm fileLayoutAlgorithm, Int32 customPESectionAlignment, Boolean verifyTypeAndFieldLayout) in ILCompiler.ReadyToRun.dll:token 0x6000131+0x127
   at ILCompiler.ReadyToRunCodegenCompilationBuilder.ToCompilation() in ILCompiler.ReadyToRun.dll:token 0x6000155+0x37
   at ILCompiler.Program.RunSingleCompilation(Dictionary`2 inFilePaths, InstructionSetSupport instructionSetSupport, String compositeRootPath, Dictionary`2 unrootedInputFilePaths, HashSet`1 versionBubbleModulesHash, CompilerTypeSystemContext typeSystemContext) in crossgen2.dll:token 0x6000165+0x627
   at ILCompiler.Program.Run(String[] args) in crossgen2.dll:token 0x6000164+0x50e
   at ILCompiler.Program.Main(String[] args) in crossgen2.dll:token 0x600016c+0x5
```

Reply:
Can you show command, which you use to invoke crossgen2? It seems that NullReferenceException should not occur if all references are passed via `-r` option

Reply:
> We should also pass: `typeSystemContext.InputFilePaths = singleCompilationInputFilePaths;`
> in other case we will have NRE:

It is not clear why InputModule was not added to `_simpleNameHashtable` inside `GetModuleFromPath` at
https://github.com/dotnet/runtime/blob/f41b259d779466068d5f7b1a16608e74e26a5e46/src/coreclr/tools/aot/crossgen2/Program.cs#L544-L556

Generally changes looks ok now.

Reply:
@gbalykov It is possible. But I think NullReferenceException should not be produced in this case. We can't find a way to understand how to fix references in this case. I have a  stable repro for that, and if you could advise a more convenient fix, I could try to do it.

Reply:
@gbalykov, I think I could rewrite my command to avoid this issue. But NRE will still be there in this case.

Reply:
@mfilippov what I mean is that NRE actually means that references miss some required files. With your changes NRE will not happen, but InputFilePaths and ReferenceFilePaths won't contain required file anyway (see `GetModuleForSimpleName`), and exception below will be thrown if `throwIfNotFound` is set. However, if `throwIfNotFound` is not set, crossgen2 will continue its execution. So, I agree that code is more correct with your changes if `--single-file-compilation` is used and follows the same logic as without `--single-file-compilation`.

Also, if this helps, you can add some log in `GetModuleForSimpleName` to check what file is not found and verify that it is passed with `-r` option.